web/src/main/webapp/table.ts

```diff
@@ -772,12 +772,38 @@ export class TableView extends RemoteObject
             return "?";  // TODO
     }
 
+    public drawDataRange(canvas: HTMLCanvasElement, position: number, count: number) : void {
+        var ctx = canvas.getContext('2d');
+
+        // A little hack to make canvas look prettier on retina displays.
+        // (Source: https://coderwall.com/p/vmkk6a/how-to-make-the-canvas-not-look-like-crap-on-retina)
+        ctx.scale(2, 2);
+        canvas.style.width = '100px';
+        canvas.style.height = '20px';
+        canvas.width = 2 * 100;
+        canvas.height = 2 * 20;
+
+        let x = canvas.width * position / this.rowCount;
+        let width = Math.max(1, canvas.width * count / this.rowCount);
+        ctx.fillStyle = 'rgb(200, 200, 200)';
+        ctx.fillRect(x, 0, width, canvas.height);
+
+        ctx.textBaseline = "middle";
+        ctx.textAlign = "right";
+        ctx.font = '36px sans';
+        ctx.fillStyle = 'black';
+        ctx.fillText(significantDigits(position), canvas.width, canvas.height / 2);
+    }
+
     public addRow(row : RowView, cds: ColumnDescription[]) : void {
         let trow = this.tBody.insertRow();
 
+        let position = this.startPosition + this.dataRowsDisplayed;
+
         let cell = trow.insertCell(0);
-        cell.style.textAlign = "right";
-        cell.textContent = significantDigits(this.startPosition + this.dataRowsDisplayed);
+        let canvas = document.createElement("canvas");
```

Comment:
I think you should actually create the canvas inside the drawDataRange function; in this way, how you draw it is entirely hidden in that function.